### PR TITLE
Add logging system callback

### PR DESCRIPTION
**Pull request by zorowk**

### CODE REVIEW

File: src/util/qwlogging.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// Copyright (C) 2022 WenHao Peng <near.kingzero@gmail.com>.
```
Comment (wineee) [inline src/util/qwlogging.cpp:1]:
```suggestion
// Copyright (C) 2024 WenHao Peng <near.kingzero@gmail.com>.
```

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/util/qwlogging.cpp (hunk 1 above)
Comment (zorowk) [reply src/util/qwlogging.cpp]: fix

File: src/util/qwlogging.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// Copyright (C) 2022 WenHao Peng <near.kingzero@gmail.com>.
```
Comment (wineee) [inline src/util/qwlogging.h:1]:
```suggestion
// Copyright (C) 2024 WenHao Peng <near.kingzero@gmail.com>.
```
same

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/util/qwlogging.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+#include "qwlogging.h"
+
+#include <QString>
+#include <QDebug>
+#include <QLoggingCategory>
+
+QW_BEGIN_NAMESPACE
+
+Q_LOGGING_CATEGORY(QWlroots, "qt.qwlroots");
```
Comment (zccrs) [inline src/util/qwlogging.cpp:12]: No `qt.`

File: src/util/qwlogging.h [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+#pragma once
+
+#include <qwglobal.h>
+extern "C" {
+#include <wlr/util/log.h>
+}
+
+QW_BEGIN_NAMESPACE
+
+void QW_EXPORT qw_log_callback(enum wlr_log_importance importance, const char *fmt, va_list args);
```
Comment (zccrs) [inline src/util/qwlogging.h:13]: How to use?

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/util/qwlogging.h (hunk 4 above)
Comment (zorowk) [reply src/util/qwlogging.h:13]: `WServer::WServer(WServerPrivate &dd, QObject *parent)
    : QObject(parent)
    , WObject(dd)
{
#ifdef QT_DEBUG
    wlr_log_init(WLR_DEBUG, qw_log_callback);
#else
    wlr_log_init(WLR_INFO, qw_log_callback);
#endif
}
`   waylib will usage this

Comment (zccrs) [reply src/util/qwlogging.h:13]: No, you should export a `QWLog` class, and it's has a `init` static function.
We can use `qw_log_callback` function on default, but this function don't export to user.

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/util/qwlogging.cpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+// SPDX-License-Identifier: Apache-2.0 OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+#include "qwlogging.h"
+
+#include <QDebug>
+#include <QLoggingCategory>
+
+QW_BEGIN_NAMESPACE
+
+Q_LOGGING_CATEGORY(QWlroots, "qwlroots");
```
Comment (justforlxz) [inline src/util/qwlogging.cpp:11]: 这里分类应该是 wlroots 吧。 @zccrs 你觉得呢？

File: src/util/qwlogging.cpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+void QWLog::init()
+{
+#ifdef QT_DEBUG
+    wlr_log_init(WLR_DEBUG, log_callback);
```
Comment (zccrs) [inline src/util/qwlogging.cpp:16]: Always enable debug here. We can control default log level via Q_LOGGING_CATEGORY.

File: src/util/qwlogging.h [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+QW_BEGIN_NAMESPACE
+
+class QW_EXPORT QWLog {
+public:
+    static void init();
+
+private:
+    static void log_callback(enum wlr_log_importance importance, const char *fmt, va_list args);
```
Comment (zccrs) [inline src/util/qwlogging.h:18]: Dont't in header file.

File: src/util/qwlogging.h [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+#include <qwglobal.h>
+extern "C" {
+#include <wlr/util/log.h>
+}
+
+QW_BEGIN_NAMESPACE
+
+class QW_EXPORT QWLog {
+public:
+    static void init();
```
Comment (zccrs) [inline src/util/qwlogging.h:15]: Needs `QtMsgType` argument.

Comment (zccrs) [reply src/util/qwlogging.h:15]: And add an override function `static void init(wlr_log_importance, callbackFunction)`

File: src/util/qwlogging.cpp (hunk 5 above)
Comment (zccrs) [reply src/util/qwlogging.cpp:11]: Yes

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/util/qwlogging.h (hunk 8 above)
Comment (zorowk) [reply src/util/qwlogging.h:15]: you mean  wlr_log_importance  There is no way QtMsgType can be used here

**Review by zccrs — APPROVED** [review]

**Pull request merged**